### DISCUSSION TITLE
azure devops hosted agents + match = ignoring certificates from match and trying to create new

### DISCUSSION THREAD

**Discussion by defaultbr** [fastlane/fastlane#17826] [General]
Hello guys, i'm trying to implement the match + bucket s3 into a ci system, but match is only downloading the certificates, decrypting it but not installing into the keychain and than complaining that it should create another one, why?

the output:

![image](https://user-images.githubusercontent.com/5432351/102551945-1c144400-409f-11eb-9d13-306231dc48de.png)

> 
> DEBUG [2020-12-17 22:28:55.30]: Downloading file from S3 'README.md' on bucket ios-certificates
> DEBUG [2020-12-17 22:28:55.36]: Downloading file from S3 'certs/development/R724DZ7SK8.cer' on bucket ios-certificates
> .....
> DEBUG [2020-12-17 22:28:55.99]: Successfully downloaded files from S3 to /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20201217-966-c40rzk
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted 'R7X4XZ7SK8.cer'
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted 'R7X4XZ7SK8.p12'
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted '55XXXJZRK9.p12'
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted '55XXXJZRK9.cer'
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted 'AppStore_***.mobileprovision'
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted 'Development_***.dev.mobileprovision'
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted 'Development_br.com.m.MCs.mobileprovision'
> INFO [2020-12-17 22:28:56.00]: 🔓  Decrypted 'AdHoc_***.qa.mobileprovision'
> INFO [2020-12-17 22:28:56.00]: 🔓  Successfully decrypted certificates repo
> WARN [2020-12-17 22:28:56.00]: Couldn't find a valid code signing identity for distribution... creating one for you now
> WARN [2020-12-17 22:28:56.00]: Lane Context:
> INFO [2020-12-17 22:28:56.00]: {:DEFAULT_PLATFORM=>:ios, :PLATFORM_NAME=>:ios, :LANE_NAME=>"ios certificates", :KEYCHAIN_PATH=>"~/Library/Keychains/temporary_keychain", :ORIGINAL_DEFAULT_KEYCHAIN=>"\"/Users/runner/Library/Keychains/login.keychain-db\""}
> ERROR [2020-12-17 22:28:56.00]: No code signing identity found and can not create a new one because you enabled `readonly`
> INFO [2020-12-17 22:28:56.00]: Successfully generated documentation at path '/Users/runner/work/1/s/fastlane/README.md'
> 

ERROR:

 **No code signing identity found and can not create a new one because you enabled `readonly`**
 
 I don't want to create new certificate, i want this certificates from the bucket
 

the lanes:

```
desc "load certificates"
lane :certificates do |options|
  if options[:ci]
    create_temporary_keychain
  end

  readonly = (options[:refresh_certificates] ? false : true)
  force_for_new_devices = !readonly

  install_certificate(identifier: ENV["APP_IDENTIFIER_QA"], type: 'adhoc', readonly: readonly)
  install_certificate(identifier: ENV["APP_IDENTIFIER"], type: 'appstore', readonly: readonly)

  if options[:ci]
    sh "security set-key-partition-list -S apple-tool:,apple:,codesign: -s -k #{ENV["KEYCHAIN_NAME"]} #{ENV["KEYCHAIN_PASSWORD"]}"
  end
end

private_lane :create_temporary_keychain do
  keychain_name = "temporary_keychain"
  ENV["KEYCHAIN_NAME"] = keychain_name
  ENV["KEYCHAIN_PASSWORD"] = keychain_name
  ENV["MATCH_KEYCHAIN_NAME"] = keychain_name 
  ENV["MATCH_KEYCHAIN_PASSWORD"] = keychain_name

  create_keychain(
    default_keychain: true,
    unlock: true,
    timeout: 3600,
    add_to_search_list: true
  )
end

lane :install_certificate do |options|
  match(
    team_id: ENV["TEAM_ID"],
    storage_mode: "s3",
    s3_bucket: "ios-certificates",
    s3_region: "us-east-1",
    s3_access_key: "{hidden}",
    s3_secret_access_key: "{hidden}",
    type: options[:type],
    app_identifier: options[:identifier],
    readonly: options[:readonly],
    verbose: true
  )
end
```

the fastlane env:

<details><summary>FASTLANE ENV</summary>

### Stack

| Key                         | Value                                            |
| --------------------------- | ------------------------------------------------ |
| OS                          | 10.15.7                                          |
| Ruby                        | 2.7.2                                            |
| Bundler?                    | false                                            |
| Git                         | git version 2.29.2                               |
| Installation Source         | /usr/local/lib/ruby/gems/2.7.0/bin/fastlane      |
| Host                        | Mac OS X 10.15.7 (19H15)                         |
| Ruby Lib Dir                | /usr/local/Cellar/ruby/2.7.2/lib                 |
| OpenSSL Version             | OpenSSL 1.1.1g  21 Apr 2020                      |
| Is contained                | false                                            |
| Is homebrew                 | false                                            |
| Is installed via Fabric.app | false                                            |
| Xcode Path                  | /Applications/Xcode_12.2.app/Contents/Developer/ |
| Xcode Version               | 12.2                                             |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | âœ… |
| LC_ALL   | en_US.UTF-8 | âœ… |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# Customise this file, documentation can be found here:
# https://github.com/fastlane/fastlane/tree/master/fastlane/docs
# All available actions: https://docs.fastlane.tools/actions
# can also be listed using the `fastlane actions` command

# Change the syntax highlighting to Ruby
# All lines starting with a # are ignored when running `fastlane`

# If you want to automatically update fastlane if a new version is available:
# update_fastlane

# This is the minimum version number required.
# Update this, if you use features of a newer version
require 'pathname'

fastlane_version "2.141.0"
default_platform :ios

# Utils
import("src/utils.rb")

# PLATAFORM ANDROID
import("src/android.rb")

# PLATAFORM IOS
import("src/ios.rb")
```
</details>

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.170.0 | âœ… Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                             | Version | Update-Status |
| ---------------------------------- | ------- | ------------- |
| fastlane-plugin-appcenter          | 1.11.0  | âœ… Up-To-Date  |
| fastlane-plugin-changelog          | 0.15.0  | âœ… Up-To-Date  |
| fastlane-plugin-get_version_name   | 0.2.2   | âœ… Up-To-Date  |
| fastlane-plugin-versioning_android | 0.1.0   | âœ… Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                                | Version      |
| ---------------------------------- | ------------ |
| did_you_mean                       | 1.4.0        |
| slack-notifier                     | 2.3.2        |
| atomos                             | 0.1.3        |
| CFPropertyList                     | 3.0.3        |
| claide                             | 1.0.3        |
| colored2                           | 3.1.2        |
| nanaimo                            | 0.3.0        |
| xcodeproj                          | 1.19.0       |
| rouge                              | 2.0.7        |
| xcpretty                           | 0.3.0        |
| terminal-notifier                  | 2.0.0        |
| public_suffix                      | 4.0.6        |
| addressable                        | 2.7.0        |
| multipart-post                     | 2.0.0        |
| word_wrap                          | 1.0.0        |
| tty-screen                         | 0.8.1        |
| tty-cursor                         | 0.7.1        |
| tty-spinner                        | 0.9.3        |
| babosa                             | 1.0.4        |
| colored                            | 1.2          |
| highline                           | 1.7.10       |
| commander-fastlane                 | 4.4.6        |
| excon                              | 0.78.1       |
| unf_ext                            | 0.0.7.7      |
| unf                                | 0.1.4        |
| domain_name                        | 0.5.20190701 |
| http-cookie                        | 1.0.3        |
| faraday-cookie_jar                 | 0.0.7        |
| ruby2_keywords                     | 0.0.2        |
| faraday                            | 1.1.0        |
| faraday_middleware                 | 1.0.0        |
| fastimage                          | 2.2.0        |
| gh_inspector                       | 1.1.3        |
| mini_magick                        | 4.11.0       |
| rubyzip                            | 2.3.0        |
| security                           | 0.1.3        |
| xcpretty-travis-formatter          | 1.0.0        |
| dotenv                             | 2.7.6        |
| bundler                            | 2.2.0        |
| naturally                          | 2.2.0        |
| simctl                             | 1.6.8        |
| jwt                                | 2.2.2        |
| uber                               | 0.1.0        |
| declarative                        | 0.0.20       |
| declarative-option                 | 0.1.0        |
| representable                      | 3.0.4        |
| retriable                          | 3.1.2        |
| mini_mime                          | 1.0.2        |
| multi_json                         | 1.15.0       |
| signet                             | 0.14.0       |
| memoist                            | 0.16.2       |
| os                                 | 1.1.1        |
| googleauth                         | 0.14.0       |
| httpclient                         | 2.8.3        |
| google-api-client                  | 0.38.0       |
| google-cloud-env                   | 1.4.0        |
| google-cloud-errors                | 1.0.1        |
| google-cloud-core                  | 1.5.0        |
| rake                               | 12.3.3       |
| digest-crc                         | 0.6.2        |
| google-cloud-storage               | 1.29.1       |
| emoji_regex                        | 3.2.1        |
| jmespath                           | 1.4.0        |
| aws-partitions                     | 1.405.0      |
| aws-eventstream                    | 1.1.0        |
| aws-sigv4                          | 1.2.2        |
| aws-sdk-core                       | 3.110.0      |
| aws-sdk-kms                        | 1.39.0       |
| aws-sdk-s3                         | 1.86.1       |
| uri                                | 0.10.0       |
| unicode-display_width              | 1.7.0        |
| terminal-table                     | 1.8.0        |
| plist                              | 3.5.0        |
| json                               | 2.3.0        |
| forwardable                        | 1.3.1        |
| logger                             | 1.4.2        |
| cgi                                | 0.1.0        |
| timeout                            | 0.1.0        |
| stringio                           | 0.1.0        |
| ipaddr                             | 1.2.2        |
| openssl                            | 2.1.2        |
| ostruct                            | 0.2.0        |
| strscan                            | 1.0.3        |
| date                               | 3.0.0        |
| delegate                           | 0.1.0        |
| fileutils                          | 1.4.1        |
| zlib                               | 1.1.0        |
| singleton                          | 0.1.0        |
| mini_portile2                      | 2.1.0        |
| nokogiri                           | 1.6.8.1      |
| racc                               | 1.4.16       |
| rexml                              | 3.2.3        |
| open3                              | 0.1.0        |
| yaml                               | 0.1.0        |
| psych                              | 3.1.0        |
| mutex_m                            | 0.1.0        |
| webrick                            | 1.6.0        |
| etc                                | 1.1.0        |
| fastlane-plugin-appcenter          | 1.11.0       |
| csv                                | 3.1.2        |
| fastlane-plugin-changelog          | 0.15.0       |
| fastlane-plugin-get_version_name   | 0.2.2        |
| fastlane-plugin-versioning_android | 0.1.0        |
</details>

Command at the ci pipeline: `fastlane ios certificates ci:true`

**Comment by rogerluan** [comment]
Hey @defaultbr 👋 

I'm not sure if this makes sense, but can you try giving a name to your keychain that ends with the `.keychain` extension? Such as `keychain_name = "temporary_keychain.keychain"

If that fixes the issue for you, I can go ahead and fix this in _fastlane_ 👍 

Also, if the problem persists, could you run your command using `--verbose` option, and posting the result here? 🙏

**Comment by defaultbr** [comment]
@rogerluan unfortunately did not worked =(

Same error:

`No code signing identity found and can not create a new one because you enabled `readonly`

**Reply by rogerluan** [reply]
I investigated this a little further but couldn't find anything obviously broken 😕 It seems to me that S3 is downloading the certs into a given path, and match is looking for the certs in a different path. I can't tell by the logs you provided though, I'd need to add extra logging for you to run, or reproduce the problem locally. Also, it looks like you trimmed the logs so I'm a bit confused. Eg..:

```
DEBUG [2020-12-17 22:28:55.36]: Downloading file from S3 'certs/development/R724DZ7SK8.cer' on bucket ios-certificates
.....
```

What was below that line (instead of `.....`)? Because that file above seems to be associated with a development certificate, and you're trying to install/use a `distribution` certificate (adhoc/appstore). That may be the cause of the mismatch, but I can't tell if the other cert you're installing (`55XXXJZRK9`) is a `distribution` one, because that part of the log was omitted 😅

**Reply by defaultbr** [reply]
```
WARN [2020-12-21 15:05:28.08]: from your Matchfile and Fastfile
INFO [2020-12-21 15:05:28.08]: The above is just a warning, fastlane will continue as usual now...
DEBUG [2020-12-21 15:05:28.24]: Downloading file from S3 'README.md' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.31]: Downloading file from S3 'certs/development/R724DZ7SK8.cer' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.42]: Downloading file from S3 'certs/development/R724DZ7SK8.p12' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.53]: Downloading file from S3 'certs/distribution/55WBXJZRK9.cer' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.63]: Downloading file from S3 'certs/distribution/55WBXJZRK9.p12' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.76]: Downloading file from S3 'match_version.txt' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.82]: Downloading file from S3 'profiles/adhoc/AdHoc_***.qa.mobileprovision' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.91]: Downloading file from S3 'profiles/appstore/AppStore_***.mobileprovision' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:28.99]: Downloading file from S3 'profiles/development/Development_br.com.midway.MidwayCartoes.mobileprovision' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:29.07]: Downloading file from S3 'profiles/development/Development_***.dev.mobileprovision' on bucket midway-app-ios-certificates
DEBUG [2020-12-21 15:05:29.15]: Successfully downloaded files from S3 to /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20201221-1163-kgizp
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted 'R724DZ7SK8.cer'
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted 'R724DZ7SK8.p12'
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted '55WXXJZRK9.p12'
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted '55WXXJZRK9.cer'
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted 'AppStore_***.mobileprovision'
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted 'Development_***.dev.mobileprovision'
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted 'Development_br.com.midway.MidwayCartoes.mobileprovision'
INFO [2020-12-21 15:05:29.15]: 🔓  Decrypted 'AdHoc_***.qa.mobileprovision'
INFO [2020-12-21 15:05:29.15]: 🔓  Successfully decrypted certificates repo
WARN [2020-12-21 15:05:29.15]: Couldn't find a valid code signing identity for distribution... creating one for you now
WARN [2020-12-21 15:05:29.16]: Lane Context:
INFO [2020-12-21 15:05:29.16]: {:DEFAULT_PLATFORM=>:ios, :PLATFORM_NAME=>:ios, :LANE_NAME=>"ios build_qa", :KEYCHAIN_PATH=>"~/Library/Keychains/temp_chain"}
ERROR [2020-12-21 15:05:29.16]: No code signing identity found and can not create a new one because you enabled `readonly`
```

**Reply by rogerluan** [reply]
Ah I see that there are `distribution` certs 🤔 

@defaultbr I created a branch that has better logging for us to identify where the files are being saved to and read from.

To test this branch, please modify your Gemfile as:

```ruby
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "rogerluan-match-s3-storage"
```

And run `bundle install` to apply the changes. 

Then run your command with `--verbose` are you were already doing, please 🙏 

Then post the full output here so I can analyze it 👍

**Reply by defaultbr** [reply]
I did, but i can't see anything different or extra logs, not difference, my Gemfile at the project root:

```
source "https://rubygems.org"
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "rogerluan-match-s3-storage"
```

![image](https://user-images.githubusercontent.com/5432351/102911882-70327600-445b-11eb-85b7-e3dfb01c1c0d.png)

**Reply by defaultbr** [reply]
DEBUG [2020-12-22 16:32:23.53]: Successfully downloaded files from S3 to /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20201222-5564-2ctrty


lol, why in this folder? impossible to do some kind of ls :O

**Reply by rogerluan** [reply]
🤯 that's weird 😅 here are the places I added the logs:

<img width="936" alt="image" src="https://user-images.githubusercontent.com/8419048/102991364-d1a82280-44f7-11eb-8df1-3cd32698bea0.png">

>lol, why in this folder? 

When we download them, we save them into a temp dir: `self.working_directory = Dir.mktmpdir` (Ruby creates that folder that looks weird 😅 ).

Could you check if your logs show the edits I made (screenshot above)? 😬 it should be working, I'm so confused 😅 🙈
